include/deal.II/matrix_free/portable_matrix_free.templates.h

```diff
@@ -298,14 +309,15 @@ namespace Portable
         }
 
       // Copy the data to the device
-      Kokkos::deep_copy(data->constraint_mask[color], constraint_mask_host);
-      Kokkos::deep_copy(data->local_to_global[color], local_to_global_host);
-      if (update_flags & update_quadrature_points)
+      Kokkos::deep_copy(dof_data.constraint_mask[color], constraint_mask_host);
+      Kokkos::deep_copy(dof_data.local_to_global[color], local_to_global_host);
+      if (update_flags & update_quadrature_points && dof_handler_index == 0)
         Kokkos::deep_copy(data->q_points[color], q_points_host);
+
       if (update_flags & update_JxW_values)
-        Kokkos::deep_copy(data->JxW[color], JxW_host);
+        Kokkos::deep_copy(dof_data.JxW[color], JxW_host);
       if (update_flags & update_gradients)
-        Kokkos::deep_copy(data->inv_jacobian[color], inv_jacobian_host);
+        Kokkos::deep_copy(dof_data.inv_jacobian[color], inv_jacobian_host);
```

Comment:
The repeated calls to `deep_copy` without execution space instance imply a lot of overfencing. Note sure of that ever becomes visible when profiling, though.

Reply:
You want to only have a single fence at the end, right? How would I do this and I assume this is something we can do in a separate PR?

Reply:
Yes, of course we can do that elsewhere. You would do something like
```C++
      ::dealii::MemorySpace::Default::kokkos_space::execution_space exec;
      Kokkos::deep_copy(exec, dof_data.constraint_mask[color], constraint_mask_host);
      Kokkos::deep_copy(exec, dof_data.local_to_global[color], local_to_global_host);
      if (update_flags & update_quadrature_points && dof_handler_index == 0)
        Kokkos::deep_copy(exec, data->q_points[color], q_points_host);

      if (update_flags & update_JxW_values)
        Kokkos::deep_copy(exec, dof_data.JxW[color], JxW_host);
      if (update_flags & update_gradients)
        Kokkos::deep_copy(exec, dof_data.inv_jacobian[color], inv_jacobian_host);
      exec.fence();
```